teps/0156-whenexpressions-in-step.md

```diff
@@ -0,0 +1,116 @@
+---
+status: proposed
+title: WhenExpressions in Steps
+creation-date: '2024-04-15'
+last-updated: '2024-04-15'
+authors:
+- '@ericzzzzzzz'
+---
+
+# TEP-0156: WhenExpressions in Steps
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+   - [Goals](#goals)
+   - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+   - [When Expressions Evaluation](#when-expressions-evaluation)
+   - [Supported Expression Languages](#supported-expression-languages)
+- [Test Plan](#test-plan)
+- [Alternatives](#alternatives)
+
+## Summary
+This TEP proposes the addition of When Expressions to individual Steps within a Task. This will allow for fine-grained control over the execution of steps based on conditions like the status of previous step results. This enhancement will provide greater flexibility and control in defining task logic, enabling more complex and dynamic workflows.
+## Motivation
```

Comment:
Anything described here can be achieved by "relatively" smart scripts in Steps, am I right ?
The "main" difference is in one case it's expressed with the tekton API, and on the other case, it's in "a script" or "in code".

Reply:
Only for inlined steps. This allows us to control the execution of stepActions where we don't have the ability to edit the scripts.

Reply:
Yes, I can hear that 😛 But one could also "condition" this in the `StepAction` with a parameter (probably a bit convoluted)

Reply:
That is a valid point, however, one can make same argument about `when` expression in Tasks 😉. 
IMHO, declarative conditionals in a CI/CD pipeline are more readable and improve maintainability of the pipeline.

Reply:
@haroonc on the other hand, it make the "split" between local build and CI/CD bigger 😛 but that's a completely different topic/debate 👼🏼